### PR TITLE
[BlockDAO] add FileDAO interface to manage chain db file

### PR DESCRIPTION
**Pull request by dustinxie**
1. this PR only adds the FileDAO interface and move current block read/write code to filedao_legacy.go
2. next PR will handle the new file implementation
3. last PR will take care how to handle chain db file depending on sf.Height(), and starting chain only on statedb

### CODE REVIEW

File: blockindex/indexbuilder.go [hunk 1]
```diff
@@ -151,30 +151,6 @@ func (ib *IndexBuilder) init() error {
 	if startHeight >= tipHeight {
 		// successfully migrated to latest block
 		zap.L().Info("Finished migrating DB", zap.Uint64("height", startHeight))
-		return ib.purgeObsoleteIndex()
```
Comment (dustinxie) [inline blockindex/indexbuilder.go:154]: code used for index db separation (about 1 year ago), not needed now

File: blockchain/blockchain.go [hunk 2]
```diff
@@ -278,7 +278,7 @@ func (bc *blockchain) TipHash() hash.Hash256 {
 func (bc *blockchain) TipHeight() uint64 {
 	tipHeight, err := bc.dao.Height()
 	if err != nil {
-		log.L().Panic("failed to get tip height", zap.Error(err))
+		log.L().Error("failed to get tip height", zap.Error(err))
```
Comment (yutongp) [inline blockchain/blockchain.go:281]: what can we do instead of panic if we can not get tip height?

**Comment by yutongp** [comment]
I assume this diff only does two things:
1. create a new fileDao interface
2. move existing code to fileDaolegacy implemention.

if that's the case, I think the diff should be safe. but the diff is big, so anything in this diff which is not 1 and 2, i don't know if that part of code is safe or not.

File: blockchain/blockchain.go (hunk 2 above)
Comment (dustinxie) [reply blockchain/blockchain.go]: changed it back

there is a wrong e2etest, it is root-caused and fixed

**Comment by dustinxie** [comment]
> I assume this diff only does two things:
> 
> 1. create a new fileDao interface
> 2. move existing code to fileDaolegacy implemention.
> 
> if that's the case, I think the diff should be safe. but the diff is big, so anything in this diff which is not 1 and 2, i don't know if that part of code is safe or not.

that is right, the diff has only 1 and 2

File: server/itx/server.go [hunk 3]
```diff
@@ -221,6 +217,11 @@ func StartServer(ctx context.Context, svr *Server, probeSvr *probe.Server, cfg c
 	if err := adminserv.Shutdown(ctx); err != nil {
 		log.L().Error("Error when serving metrics data.", zap.Error(err))
 	}
+	if cfg.System.HeartbeatInterval > 0 {
+		if err := task.Stop(ctx); err != nil {
+			log.L().Panic("Failed to stop heartbeat routine.", zap.Error(err))
+		}
+	}
```
Comment (dustinxie) [inline server/itx/server.go:224]: Stop() should be called in the LIFO order of Start()

task starts after svr (and used svr), so should stop before svr

File: e2etest/local_transfer_test.go [hunk 4]
```diff
@@ -312,28 +311,29 @@ func TestLocalTransfer(t *testing.T) {
...
 	defer func() {
-		time.Sleep(10 * time.Second)
 		require.NoError(probeSvr.Stop(ctx))
-		require.NoError(svr.Stop(ctx))
 	}()
 
+	// Start server
+	ctx, cancel := context.WithTimeout(ctx, 32*time.Second)
+	defer cancel()
+	go itx.StartServer(ctx, svr, probeSvr, cfg)
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:328]: should use a timeout context, which calls svr.Stop() when done

currently, the svr is never stopped

File: e2etest/local_transfer_test.go [hunk 5]
```diff
@@ -312,28 +311,29 @@ func TestLocalTransfer(t *testing.T) {
...
+
 	// target address for grpc connection. Default is "127.0.0.1:14014"
 	grpcAddr := fmt.Sprintf("127.0.0.1:%d", apiPort)
 	grpcctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
 	conn, err := grpc.DialContext(grpcctx, grpcAddr, grpc.WithBlock(), grpc.WithInsecure())
 	require.NoError(err)
+	defer conn.Close()
```
Comment (dustinxie) [inline e2etest/local_transfer_test.go:336]: close the connection

File: server/itx/server.go [hunk 6]
```diff
@@ -174,6 +174,11 @@ func StartServer(ctx context.Context, svr *Server, probeSvr *probe.Server, cfg c
 		log.L().Fatal("Failed to start server.", zap.Error(err))
 		return
 	}
+	defer func() {
+		if err := svr.Stop(ctx); err != nil {
+			log.L().Panic("Failed to stop server.", zap.Error(err))
+		}
+	}()
```
Comment (dustinxie) [inline server/itx/server.go:181]: stop() should be called in the LIFO order of start(), so use defer to do that

Comment (yutongp) [reply server/itx/server.go:181]: Maybe change task to not to use defer to make it more clear? Just stop them in order at end of the function

Comment (dustinxie) [reply server/itx/server.go:181]: i think this way is also clear?

```
a.Start()
defer a.Stop()

b.Start()
defer b.Stop()

c.Start()
defer c.Stop()

<- channel to receive exit or cancel signal
```
then we don't need to worry about the exact order. defer is exactly designed to handle this kind of usage.

**Comment by dustinxie** [comment]
with the upcoming change, BlockDAO will handle multiple KVStore, so remove db.KVStore input from NewBlockDAO(), let FileDAO handle that

**Pull request merged**